File: frontend/src/types/response.ts
import { BaseEntity, ID } from './common';
import { FormData, FormSubmissionStatus } from './form';

export interface Response extends BaseEntity {
  formId: ID;
  data: FormData;
  referenceCode: string;
  status: ResponseStatus;
  submittedAt?: string;
  reviewedAt?: string;
  reviewedBy?: ID;
  notes?: string;
  tags?: string[];
  priority?: ResponsePriority;
}

export type ResponseStatus = FormSubmissionStatus;

export type ResponsePriority = 'low' | 'medium' | 'high' | 'urgent';

export interface ResponseListItem {
  id: ID;
  formId: ID;
  referenceCode: string;
  status: ResponseStatus;
  priority?: ResponsePriority;
  submittedAt: string;
  createdAt: string;
}

export interface CreateResponsePayload {
  formId: ID;
  data: FormData;
  status?: ResponseStatus;
}

export interface UpdateResponsePayload {
  data?: FormData;
  status?: ResponseStatus;
  notes?: string;
  tags?: string[];
  priority?: ResponsePriority;
}

export interface ResponseFilter {
  status?: ResponseStatus;
  priority?: ResponsePriority;
  formId?: ID;
  dateFrom?: string;
  dateTo?: string;
  tags?: string[];
  search?: string;
}

export interface ResponseStats {
  total: number;
  byStatus: Record<ResponseStatus, number>;
  byPriority: Record<ResponsePriority, number>;
  recentCount: number;
}

export interface ResponseAnalytics {
  totalResponses: number;
  responsesByCategory: Record<string, number>;
  responsesByLocation: Record<string, number>;
  responsesByTime: {
    date: string;
    count: number;
  }[];
  averageResponseTime?: number;
}

export default {
  Response
};

File: backend/app/schemas/response.py
from pydantic import Field
from typing import Optional, Dict, Any, List
from datetime import datetime
from uuid import UUID
from app.schemas.common import BaseSchema, TimestampSchema, IDSchema
from app.models.response import ResponseStatus, ResponsePriority


class ResponseBase(BaseSchema):
    form_id: UUID
    data: Dict[str, Any] = Field(default_factory=dict)


class ResponseCreate(ResponseBase):
    status: Optional[ResponseStatus] = ResponseStatus.SUBMITTED


class ResponseUpdate(BaseSchema):
    data: Optional[Dict[str, Any]] = None
    status: Optional[ResponseStatus] = None
    notes: Optional[str] = None
    tags: Optional[List[str]] = None
    priority: Optional[ResponsePriority] = None


class ResponseResponse(ResponseBase, IDSchema, TimestampSchema):
    reference_code: str
    status: ResponseStatus
    priority: Optional[ResponsePriority] = None
    notes: Optional[str] = None
    tags: Optional[List[str]] = None
    submitted_at: Optional[datetime] = None
    reviewed_at: Optional[datetime] = None
    reviewed_by: Optional[UUID] = None
    metadata: Optional[Dict[str, Any]] = None


class ResponseListItem(BaseSchema):
    id: UUID
    form_id: UUID
    reference_code: str
    status: ResponseStatus
    priority: Optional[ResponsePriority] = None
    submitted_at: Optional[datetime] = None
    created_at: datetime


class ResponseFilter(BaseSchema):
    status: Optional[ResponseStatus] = None
    priority: Optional[ResponsePriority] = None
    form_id: Optional[UUID] = None
    date_from: Optional[datetime] = None
    date_to: Optional[datetime] = None
    search: Optional[str] = None